source/portable/NetworkInterface/MPS3_AN552/NetworkInterface.c

```diff
@@ -0,0 +1,367 @@
+    ARM_ETH_LINK_STATE eEthernetLinkState;
+    uint32_t ulDataRead;
+
+    ( void ) pvParameters;
+
+    for( ; ; )
+    {
+        /* Wait for the Ethernet ISR to receive a packet or timeout (100ms). */
+        ulTaskNotifyTake( pdFALSE, xBlockTime );
+
+        ulDataRead = prvLowLevelInput( &pxNetworkBuffer );
+
+        if( ulDataRead > 0 )
+        {
+            xRxEvent.pvData = ( void * ) pxNetworkBuffer;
+
+            pxNetworkBuffer->pxInterface = pxMyInterface;
+            pxNetworkBuffer->pxEndPoint = FreeRTOS_MatchingEndpoint( pxMyInterface, pxNetworkBuffer->pucEthernetBuffer );
+
+            if( xSendEventStructToIPTask( &xRxEvent, ( TickType_t ) 0 ) == pdFAIL )
+            {
+                vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
+            }
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static uint32_t prvLowLevelInput( NetworkBufferDescriptor_t ** pxNetworkBuffer )
+{
+    const TickType_t xDescriptorWaitTime = pdMS_TO_TICKS( 0 );
+    uint32_t ulMessageLength = 0;
+    int32_t lReceivedBytes = 0;
+
+    ulMessageLength = xEthernetInterface0.pxEthernetMACDriver->GetRxFrameSize();
+
+    if( ulMessageLength != 0 )
+    {
+        *pxNetworkBuffer = pxGetNetworkBufferWithDescriptor( ulMessageLength,
+                                                             xDescriptorWaitTime );
+
+        if( *pxNetworkBuffer != NULL )
+        {
+            lReceivedBytes = xEthernetInterface0.pxEthernetMACDriver->ReadFrame(
+                ( ( *pxNetworkBuffer )->pucEthernetBuffer ),
+                ulMessageLength );
+
+            if( lReceivedBytes >= 0 )
+            {
+                ( *pxNetworkBuffer )->xDataLength = lReceivedBytes;
+            }
+            else
+            {
+                FreeRTOS_debug_printf( ( "Error: Ethernet driver ReadFrame returned %d \n", lReceivedBytes ) );
+                lReceivedBytes = 0;
+            }
+        }
+        else
+        {
+            FreeRTOS_debug_printf( ( "pxNetworkBuffer = NULL\n" ) );
+            /* No memory available for NetworkBuffer, drop the frame */
+            xEthernetInterface0.pxEthernetMACDriver->ReadFrame(
+                NULL,
+                0 );
+        }
+    }
+
+    return lReceivedBytes;
+}
+/*-----------------------------------------------------------*/
+
+static BaseType_t xLAN91C111_NetworkInterfaceInitialise( NetworkInterface_t * pxInterface )
+{
+    BaseType_t xReturn = pdPASS;
+    UBaseType_t eEthernetLinkStateTimeOutCount = 0;
+    ARM_ETH_MAC_CAPABILITIES xEthernetMACCapabilities;
+    ARM_ETH_LINK_STATE eEthernetLinkState = ARM_ETH_LINK_DOWN;
+    ARM_ETH_LINK_INFO xEthernetLinkInfo;
+    uint32_t ulMACControlArg;
+
+    ( void ) pxInterface;
+
+    if( xRxTaskHandle == NULL )
+    {
+        /* Task that reads incoming Ethernet frames and sends it FreeRTOS
+         * TCP/IP stack. */
+        xReturn = xTaskCreate( prvRxTask,
```

Comment:
Creating the task should be the last step in the initialisation, because `prvRxTask` takes over the ownership of the EMAC. It can be disturbing (fatal) when the IP-task calls `NetworkInterfaceInitialise()`, working on the EMAC, and `prvRxTask` accessing DMA and the PHY.
So let `xTaskCreate()` happen in the last state of the initialisation.

Reply:
Chime in with this opinion. To avoid accessing non-init hardware/variables, we should move this to the last.

Reply:
Done. Seems like an issue in existing network interface https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/tree/main/source/portable/NetworkInterface/MPS2_AN385 which I used as a reference for this.